ta/pkcs11/src/handle.c

```diff
@@ -86,17 +89,28 @@ void *handle_put(struct handle_db *db, uint32_t handle)
 
 void *handle_lookup(struct handle_db *db, uint32_t handle)
 {
-	if (!db_and_handle_are_valid(db, handle))
+	if (!db_and_handle_are_valid(db, handle) ||
+	    db->ptrs[handle] == INVALID_HANDLE_PTR)
 		return NULL;
 
 	return db->ptrs[handle];
 }
 
+void handle_invalidate(struct handle_db *db, uint32_t handle)
+{
+	if (db_and_handle_are_valid(db, handle)) {
+		if (!db->ptrs[handle])
+			TEE_Panic(TEE_ERROR_GENERIC);
+
+		db->ptrs[handle] = INVALID_HANDLE_PTR;
```

Comment:
If I read correctly -- this will effectively mark handle ID "consumed" forever?

Eg. we have `(uint32_t)~0 - 2` handles and if they have been invalidated then game is over?

Reply:
What I see happening in our devices is that the connection could be active for months -- meaning the connection to PKCS#11 TA is open for months :)

BUT I do not expect to see too much activity. There could be new keys pairs created in limited fashion, some certs added/remove. So in real life I suppose if some one would really try to exhaust the handles then the TA's HEAP size would be the real limit. That being said HEAP size limit comes much much faster.

Reply:
Actually we could compress the list... eg. ID just going forward and then use memory for those objects that are actually available. But again -- perhaps monitor the situation first.

Reply:
The objects are destroyed so there is no problem with memory used for objects. It is only the handles that remain, and the space allocated for the client handle database which could grow and exhaust the TA heap. Franlky i see no way out but a redesign of the session object handles database.

That said, the issue reported reported by @vting-seq (https://github.com/OP-TEE/optee_os/issues/6005) is real and I think this P-R resolves it. So I would greatly prefer we merge this P-R and take time to address the heap exhaustion  you foresee. Or maybe I did not understand the actuallity of this heap issue.